test/tests/dom/shadowDOMFunctions.js

```diff
@@ -0,0 +1,199 @@
+/* @flow */
+
+import { isShadowElement, getShadowRoot, getShadowHost, insertShadowSlot } from '../../../src';
+
+// This component is needed for testing the case when shadowRoot and shadowDOM are the same
+const customWebWrapper = class extends HTMLElement {
+    constructor() {
+        super();
+        const shadowRoot = this.attachShadow({ mode: 'open' });
+        const shadowDOMContainer = document.createElement('div');
+        shadowDOMContainer.setAttribute('id', 'inner-host-div');
+        shadowRoot.appendChild(shadowDOMContainer);
+    }
+};
+
+const customWebComponent = class extends HTMLElement {
+    constructor() {
+        super();
+        const shadowHost = this.attachShadow({ mode: 'open' });
+        const shadowDOMContainer = document.createElement('div');
+        const testSpan = document.createElement('span');
+        testSpan.setAttribute('id', 'inner-span');
+        testSpan.textContent = 'text from custom web component';
+        // Append it to the shadow root
+        shadowDOMContainer.appendChild(testSpan);
+        shadowHost.appendChild(shadowDOMContainer);
+    }
+};
+
+before(() => {
```

Comment:
It seems unusual to have a `before()` block outside of the `describe` blocks. With each test, we want to make sure we have a clean DOM. That way tests are isolated from one another and will not share any expected state in the DOM. To accomplish this, can we use a `beforeEach()` instead? Here's what I'm thinking:

```js

describe('web components', () => {
    describe('isShadowElement cases', () => {
        beforeEach(() => {
            // clean the DOM
            document.body.innerHTML = '';

            // do the setup required for the specific describe block we are in (ex: some blocks do not need custom-wrapper)
            customElements.define('custom-web-component', customWebComponent);
            customElements.define('custom-wrapper', customWebWrapper);

            if (!body) {
                throw new Error('Body not found');
            }

            const customElement = document.createElement('custom-web-component');
            body.appendChild(customElement);

            if (!customElement || !customElement.shadowRoot) {
                throw new Error('custom element does not have shadow root');
            }

        });

    describe('getShadowRoot cases', () => {

    describe('getShadowHost cases', () => {
})
```

Reply:
changed! thanks Greg!